vmm/src/igvm/igvm_loader.rs

```diff
@@ -1,26 +1,445 @@
+        IGVM_VHS_MEMORY_MAP_ENTRY {
+            starting_gpa_page_number: range.base / HV_PAGE_SIZE,
+            number_of_pages: range.size as u64 / HV_PAGE_SIZE,
+            entry_type: MemoryMapEntryType::MEMORY,
+            flags: 0,
+            reserved: 0,
+        }
+    } else {
+        IGVM_VHS_MEMORY_MAP_ENTRY {
+            starting_gpa_page_number: range.base / HV_PAGE_SIZE,
+            number_of_pages: range.size as u64 / HV_PAGE_SIZE,
+            entry_type: MemoryMapEntryType::PLATFORM_RESERVED,
+            flags: 0,
+            reserved: 0,
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Copy, Clone)]
+struct GpaPages {
+    pub gpa: u64,
+    pub page_type: u32,
+    pub page_size: u32,
 }
+
 ///
 /// Load the given IGVM file to guest memory.
 /// Right now it only supports SNP based isolation
 ///
 pub fn load_igvm(
-    mut _file: &std::fs::File,
-    _memory_manager: Arc<Mutex<MemoryManager>>,
-    _cpu_manager: Arc<Mutex<CpuManager>>,
-    _mem_regions: Vec<ArchMemRegion>,
-    _proc_count: u32,
-    _cmdline: &str,
-    #[cfg(feature = "snp")] host_data: &str,
+    mut file: &std::fs::File,
+    memory_manager: Arc<Mutex<MemoryManager>>,
+    cpu_manager: Arc<Mutex<CpuManager>>,
+    mem_regions: Vec<ArchMemRegion>,
+    proc_count: u32,
+    cmdline: &str,
 ) -> Result<Box<IgvmLoadedInfo>, Error> {
-    unimplemented!()
+    let mut loaded_info: Box<IgvmLoadedInfo> = Box::new(IgvmLoadedInfo::new());
+    let command_line = CString::new(cmdline).map_err(Error::InvalidCommandLine)?;
+    let mut first_gpa: u64 = 0;
+    let mut gpa_found: bool = false;
+    let mut file_contents = Vec::new();
+    let memory = memory_manager.lock().as_ref().unwrap().guest_memory();
+    let mut gpas: Vec<GpaPages> = Vec::new();
+
+    file.seek(SeekFrom::Start(0)).map_err(Error::Igvm)?;
+    file.read_to_end(&mut file_contents).map_err(Error::Igvm)?;
+
+    let igvm_file = IgvmFile::new_from_binary(&file_contents, Some(IsolationType::Snp))
+        .map_err(Error::InvalidIgvmFile)?;
+
+    let (mask, max_vtl) = match &igvm_file.platforms()[0] {
+        IgvmPlatformHeader::SupportedPlatform(info) => {
+            debug_assert!(info.platform_type == IgvmPlatformType::SEV_SNP);
+            (info.compatibility_mask, info.highest_vtl)
+        }
+    };
+    let max_vtl = max_vtl
+        .try_into()
+        .expect("igvm file should be valid after new_from_binary");
+    let mut loader = Loader::new(memory, max_vtl);
+
+    #[derive(Debug)]
+    enum ParameterAreaState {
+        /// Parameter area has been declared via a ParameterArea header.
+        Allocated { data: Vec<u8>, max_size: u64 },
+        /// Parameter area inserted and invalid to use.
+        Inserted,
+    }
+    let mut parameter_areas: HashMap<u32, ParameterAreaState> = HashMap::new();
+
+    // Import a parameter to the given parameter area.
+    let import_parameter = |parameter_areas: &mut HashMap<u32, ParameterAreaState>,
```

Comment:
can this just be a normal function instead of defined here as a variable?

Reply:
That can be. What benefit do we get then? Readability?

Reply:
Yes.
If you put a function in a variable like this, it looks like it is a closure and needs to capture some state from the surrounding scope. But this doesn't seem to be the case here, so it's a bit confusing and unnecessary.